ring/spread_minimizing_token_generator.go

```diff
@@ -0,0 +1,279 @@
+	}
+	return strconv.Atoi(parts[3])
+}
+
+// getZoneID gets a zone name and a slice of sorted zones,
+// and return the index of the zone in the slice.
+func getZoneID(zone string, sortedZones []string) (int, error) {
+	if !slices.IsSorted(sortedZones) {
+		slices.Sort(sortedZones)
+	}
+	index := sort.SearchStrings(sortedZones, zone)
+	if index >= len(sortedZones) {
+		return 0, errorZoneNotValid(zone)
+	}
+	return index, nil
+}
+
+// generateFirstInstanceTokens calculates a set of tokens that should be assigned to the first instance (with id 0)
+// of the zone of the underlying instance.
+func (t *SpreadMinimizingTokenGenerator) generateFirstInstanceTokens() Tokens {
+	zonesCount := len(t.zones)
+	tokensPerInstance := t.cfg.TokensPerInstance
+	tokenDistance := (totalTokensCount / tokensPerInstance / zonesCount) * zonesCount
+	tokens := make(Tokens, 0, tokensPerInstance)
+	for i := 0; i < tokensPerInstance; i++ {
+		token := uint32(i*tokenDistance) + uint32(t.zoneID)
+		tokens = append(tokens, token)
+	}
+	return tokens
+}
+
+// calculateNewToken determines where in the range represented by the given ringToken should a new token be placed
+// in order to satisfy the constraint represented by the optimalTokenOwnership. This method assumes that:
+// - ringToken.token % zonesCount == ringToken.prevToken % zonesCount
+// - optimalTokenOwnership % zonesCount == 0,
+// where zonesCount is the number of zones in the ring. The caller of this function must ensure that these assumptions hold.
+func (t *SpreadMinimizingTokenGenerator) calculateNewToken(token ringToken, optimalTokenOwnership uint32) (uint32, error) {
+	zonesCount := uint32(len(t.zones))
+	if optimalTokenOwnership < zonesCount || optimalTokenOwnership%zonesCount != 0 {
+		return 0, errorMultipleOfZonesCount(zonesCount, optimalTokenOwnership, token)
+	}
+	if token.prevToken%zonesCount != token.token%zonesCount {
+		return 0, errorLowerAndUpperBoundModulo(zonesCount, optimalTokenOwnership, token)
+	}
+	ownership := getTokenDistance(token.prevToken, token.token)
+	if ownership <= int(optimalTokenOwnership) {
+		return 0, errorDistanceBetweenTokensNotBigEnough(int(optimalTokenOwnership), ownership, token)
+	}
+	maxTokenValue := (math.MaxUint32/zonesCount - 1) * zonesCount
+	offset := maxTokenValue - token.prevToken
+	if offset < optimalTokenOwnership {
+		newToken := optimalTokenOwnership - offset
+		return newToken, nil
+	}
+	return token.prevToken + optimalTokenOwnership, nil
+}
+
+// getOptimalTokenOwnership calculates the optimal ownership of the remaining currTokensCount tokens of an instance
+// having the given current instances ownership currInstanceOwnership and the given optimal instance ownership
+// optimalInstanceOwnership. The resulting token ownership must be a multiple of the number of zones.
+func (t *SpreadMinimizingTokenGenerator) getOptimalTokenOwnership(optimalInstanceOwnership, currInstanceOwnership float64, currTokensCount uint32) uint32 {
```

Comment:
Nit, but still _important_ to keep the style consistent:
```suggestion
func (t *SpreadMinimizingTokenGenerator) optimalTokenOwnership(optimalInstanceOwnership, currInstanceOwnership float64, currTokensCount uint32) uint32 {
```

[Effective Go](https://go.dev/doc/effective_go):
> it's neither idiomatic nor necessary to put Get into the getter's name